docs/design/features/MarshalDirective.md

```diff
@@ -0,0 +1,206 @@
+
+2) The runtime attempts to find a binary with the name supplied in `DllImportAttribute`.
+
+    * Discovery of the target binary is complicated and can be influenced by the [`AssemblyLoadContext`](https://docs.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext) and [`NativeLibrary`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary) classes.
+
+3) Once the binary is found and loaded into the runtime, it is queried for the expected export name. The name of the attributed function is used by default but this is configurable by the [`DllImportAttribute.EntryPoint`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.entrypoint) property.
+
+    * This process is also influenced by additional `DllImportAttribute` properties as well as by the underlying platform. For example, the [Win32 API ANSI/UNICODE convention](https://docs.microsoft.com/windows/win32/intl/conventions-for-function-prototypes) on Windows is respected and a(n) `A`/`W` suffix may be appended to the function name if it is not immediately found.
+
+4) The IL Stub is called like any .NET method, but the address of the export is passed to the generated IL Stub via a 'hidden' argument.
+
+    * **Marshal Directive**: The user's defined method would require an identical signature to that of the one typically defined, but a new argument of type `IntPtr` should be added as the first argument. The runtime would be responsible for providing the export to the user defined function.
+
+5) The IL Stub then marshals arguments as appropriate and invokes the export via the `calli` instruction.
+
+    * **Marshal Directive**: The user's defined method would cast the supplied export to an appropriate C# function pointer type and invoke the call.
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+An example of how the snippet above could be transformed is below. The example is using the proposed API in this document as well as the [C# function pointer proposal][csharp_fptr_link]. Observe points (C), the new attribute, (D), a `partial` class, and (E) the stub with additional `IntPtr` argument that will be called in lieu of a runtime generated IL Stub.
+
+``` CSharp
+[DllImportAttribute("Kernel32.dll")]
+/* C */ [MarshalDirectiveAttribute(
+        Type = MarshalDirectiveType.UserDefinedStub,
+        UserFunctionType = typeof(Stubs),
+        UserDefinedStub = nameof(QueryPerformanceCounterStub))]
+/* A */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* B */ QueryPerformanceCounter(out count);
+
+/* D */  class partial Stubs
+{
+    /* E */ public static bool QueryPerformanceCounterStub(IntPtr fptr, out long lpPerformanceCount)
```

Comment:
Shouldn't the `fptr` be of type `void*`? That seems to fit the function pointers design better. It would also make the stub unsafe which is basically a good thing (even though it doesn't matter much). I personally don't like passing pointers around as IntPtr since it means "safe" can handle them which feels wrong.

Reply:
All valid points. The proposal has been altered substantially.